File: src/main/java/com/cjw/reactivecommunityproject/common/security/config/SecurityBean.java
package com.cjw.reactivecommunityproject.common.security.config;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

@Configuration
@RequiredArgsConstructor
public class SecurityBean {
    private static final String[] ALLOW_ORIGINS_LIST = {"/rest/*"};
    private static final String[] ALLOW_METHODS_LIST = {"GET", "POST", "PUT", "DELETE", "OPTIONS"};
    private static final String[] ALLOW_HEADERS_LIST = {"Content-Type", "Authorization"};

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration configuration = new CorsConfiguration();
        configuration.setAllowedOrigins(List.of(ALLOW_ORIGINS_LIST));
        configuration.setAllowedMethods(List.of(ALLOW_METHODS_LIST));
        configuration.setAllowedHeaders(List.of(ALLOW_HEADERS_LIST));
        configuration.setAllowCredentials(false);
        configuration.setMaxAge(3600L);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", configuration);
        return source;
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}

File: src/main/kotlin/com/devcjw/reactivecommunity/auth/config/SecurityBean.kt
package com.devcjw.reactivecommunity.auth.config

import com.devcjw.reactivecommunity.auth.config.handler.RcServerAccessDeniedHandler
import com.devcjw.reactivecommunity.auth.config.handler.RcServerAuthenticationEntryPoint
import com.devcjw.reactivecommunity.auth.dao.AuthDAO
import com.devcjw.reactivecommunity.auth.model.entity.OutRestfulEntity
import jakarta.annotation.PostConstruct
import lombok.RequiredArgsConstructor
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.server.ServerAuthenticationEntryPoint
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler

@Configuration
@RequiredArgsConstructor
class SecurityBean(
    private val authDAO: AuthDAO,
) {
    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    fun rcRcServerAuthenticationEntryPoint(): ServerAuthenticationEntryPoint {
        return RcServerAuthenticationEntryPoint()
    }

    @Bean
    fun rcServerAccessDeniedHandler(): ServerAccessDeniedHandler {
        return RcServerAccessDeniedHandler()
    }

    @Bean
    fun roleMapping(): HashMap<Long, ArrayList<OutRestfulEntity>> {
        return hashMapOf()
    }

    @PostConstruct
    private fun init() {
        authDAO.selectUserLevelResource()
            .doOnNext { userLevelResources ->
                val resourcesList = userLevelResources.resources.split("|").map {
                    val (method, pattern) = it.split(",")
                    OutRestfulEntity(method, pattern)
                }
                roleMapping()[userLevelResources.levelUid] = ArrayList(resourcesList)
            }.subscribe()
    }
}